Support kubeadm upgrade with remote etcd cluster

craigtracey opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
Currently kubeadm only performs an upgrade if the etcd cluster is
colocated with the control plane node. As this is only one possible
configuration, kubeadm should support upgrades with etcd clusters
that are not local to the node.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes # https://github.com/kubernetes/kubeadm/issues/727

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```
NONE
```

craigtracey committed:
Support kubeadm upgrade with remote etcd cluster

Currently kubeadm only performs an upgrade if the etcd cluster is
colocated with the control plane node. As this is only one possible
configuration, kubeadm should support upgrades with etcd clusters
that are not local to the node.

Signed-off-by: Craig Tracey <craigtracey@gmail.com>

timothysc commented on cmd/kubeadm/app/util/etcd.go:
```diff
@@ -18,34 +18,61 @@ package util
-type EtcdCluster interface {
-	GetEtcdClusterStatus() (*clientv3.StatusResponse, error)
+// EtcdClusterInterrogator is an interface to get etcd cluster related information
+type EtcdClusterInterrogator interface {
```
I'm of the opinion we should limit the dep expansion and move to wrap and isolate by just adding a Versions f(n). 

Most of the external processing is just string comparison, so we should limit it to that and not proliferate parsing etcd deps, b/c they have been known to shift over time. 

envelop(letter) wrap it.

craigtracey replied:
Getting versions is just a byproduct of gathering status. 

Yes, I have added a check to ensure that there is no version mismatch on the etcd nodes (ie. in case etcd was partially upgraded), but the primary objective is to connect to each etcd node and ensure that it is available before attempting an upgrade. This was the original functionality that was in place, with the exception that it only verified etcd at 127.0.0.1.

timothysc replied:
I understand what you are doing.  I'm invoking campsite rules.  Every time we tread in an area, instead of just amending the simple change we need to fix a specific issue, we should strive to make it better than what it was before, otherwise, we end up with a brittle patch-ball.  

The abstraction leakage is an example.  Lets not do that, and instead create a utility function that gets the version(s) b/c that's all your code was looking for anyway.

timothysc reviewed (changes requested):
psa - @chuckha

craigtracey committed:
Further abstract etcd status

Further abstract the etcd status by creating a new EtcdStatus which will
provide the portions of the etcd status data that we care about.
Further, add a VersionsMatch function that will verify that all etcd
versions are identical.

As there were additional asks to determine whether all members are in
the same cluster, add a check for this during the member interrogation.

Signed-off-by: Craig Tracey <craigtracey@gmail.com>

chuckha commented on cmd/kubeadm/app/phases/upgrade/compute.go:
```diff
@@ -112,13 +112,17 @@ func GetAvailableUpgrades(versionGetterImpl VersionGetter, experimentalUpgradesA
 		return nil, err
 	}
 
+	if !etcdStatus.VersionsMatch() {
```
I don't think this belongs here. This function is just about getting a current state of the world.  Maybe it could go in enforceRequirements?

craigtracey replied:
I agree that this is not ideal. This is merely addressing how we contact external etcd nodes.

We have preflight checks that address etcd cluster versions, and they do not use the code that I am introducing here. In fact, they use an HTTP client to probe an ectd cluster.  I believe that we should fix those preflight checks to make use of this etcd interrogator code, but that would be expanding the scope of this PR significantly.  I am happy to open a follow-up issue here, but in this PR trying to make the change a bit more manageable.  Thoughts?

timothysc replied:
I'm ok with the failing here, b/c versions on inconsistent.

chuckha commented on cmd/kubeadm/app/phases/upgrade/compute.go:
```diff
@@ -112,13 +112,17 @@ func GetAvailableUpgrades(versionGetterImpl VersionGetter, experimentalUpgradesA
 		KubeadmVersion:  kubeadmVersionStr,
 		KubeletVersions: kubeletVersions,
-		EtcdVersion:     etcdStatus.Version,
+		EtcdVersion:     etcdStatus.Versions[0],
```
This would be more useful if it listed output like we do for kubelet. That shows the number of nodes at which version e.g.  `2 x 1.9.5, 1 x 1.9.6`

chuckha commented on cmd/kubeadm/app/cmd/upgrade/plan.go:
```diff
@@ -60,8 +61,11 @@ func RunPlan(parentFlags *cmdUpgradeFlags) error {
 
-	// Define Local Etcd cluster to be able to retrieve information
-	etcdCluster := kubeadmutil.LocalEtcdCluster{}
+	e := upgradeVars.cfg.Etcd
```
What do you think about defaulting to a local etcd if no etcd configuration is supplied?

fabriziopandini replied:
Afaik kubeadm currently is not explicit about local etcd because e.g. the code assumes that `len(etcd.endpoints)>0` is signal for external etcd (while technically also local etcd has an endpoint).
If we move to an explicit declaration of local etcd as well, IMO we should be consistent in all the kubeadm code base

stealthybox replied:
We've had discussions before about how sharing the same Etcd config struct across different Etcd implementations has caused smell throughout the code.

There really should be a more clear configuration than `len(cfg.Etcd.Endpoints)>0` for switching between External and StaticPod etcd.

chuckha reviewed (changes requested):
a few comments

timothysc commented on cmd/kubeadm/app/phases/upgrade/compute_test.go:
```diff
@@ -62,19 +65,35 @@ func (f *fakeVersionGetter) KubeletVersions() (map[string]uint16, error) {
-	client.Version = "3.1.12"
-	return client, nil
+func (f fakeEtcdCluster) GetEtcdClusterStatus() (*etcdutil.EtcdStatus, error) {
+	return &etcdutil.EtcdStatus{
```
Could probably just put this in the tests.

timothysc commented on cmd/kubeadm/app/util/etcd/etcd.go:
```diff
@@ -0,0 +1,152 @@
+// EtcdStatus provides data about an etcd cluster
+type EtcdStatus struct {
+	Versions []string
+	Members  []uint64
```
I think you could simplify the logic and extra functions by just having map[int]string

craigtracey replied:
I have added it in this manner because I think there may be a handful of other EtcdStatus interface functions that might be handy (ie. determining whether the version of etcd is at a version that is appropriate). As Members is not really being used yet, maybe we ditch that field?

xiangpengzhao reviewed (commented):
some nits and one concern.

xiangpengzhao commented:
/ok-to-test

timothysc commented:
@craigtracey - ping

detiber commented:
@craigtracey This is going to conflict with the work @stealthybox and I have in flight for fixing the etcd tls upgrade. I'll take a closer look at reconciling today.

timothysc commented:
@detiber ?

timothysc reviewed (commented):
/approve 

Needs a rebase and comments addressed before lgtm.

detiber commented:
@craigtracey I created a WIP PR to attempt to provide the meat of this PR here: https://github.com/kubernetes/kubernetes/pull/63495

timothysc commented:
closing in favor of  #63495
